text/0097-log-data-model.md

```diff
@@ -0,0 +1,1005 @@
+  custom data that the application may want to include in the logs.
+
+When designing this data model I followed the following reasoning to make a
+decision about when to use use a top-level named field:
+
+- The field needs to be either mandatory for all records or be frequently
+  present in well-known log and event formats (such as `Timestamp`) or is
+  expected to be often present in log records in upcoming logging systems (such
+  as `TraceId`).
+
+- The field’s semantics must be the same for all known log and event formats and
+  can be mapped directly and unambiguously to this data model.
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |Description
+---------------|--------------------------------------------
+Timestamp      |Time when the event occurred.
+TraceId        |Request trace id.
+SpanId         |Request span id.
+TraceFlags     |W3C trace flag.
+SeverityText   |The severity text (also known as log level).
+SeverityNumber |Numerical value of the severity.
+ShortName      |Short event identifier.
+Body           |The body of the log record.
+Resource       |Describes the source of the log.
+Attributes     |Additional information about the event.
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
+
+Type: Timestamp, uint64 nanoseconds since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
+field is optional, it may be missing the timestamp is unknown.
+
+### Trace Context Fields
+
+#### Field: `TraceId`
+
+Type: byte sequence.
+
+Description: Optional request trace id as defined in
+[W3C Trace Context](https://www.w3.org/TR/trace-context/#trace-id). Can be set
+for logs that are part of request processing and have an assigned trace id.
+
+#### Field: `SpanId`
+
+Type: byte sequence.
+
+Description: Optional span id. Can be set for logs that are part of a particular
+processing span. If SpanId is present TraceId SHOULD be also present.
+
+#### Field: `TraceFlags`
+
+Type: byte.
+
+Description: Optional trace flag as defined in
+[W3C Trace Context](https://www.w3.org/TR/trace-context/#trace-flags)
+specification. At the time of writing the specification defines one flag - the
+SAMPLED flag.
+
+### Severity Fields
+
+#### Field: `SeverityText`
```

Comment:
After reading all the rules for comparison and transforms, I am not sure I see much value in this field. For log formats that unambiguously map to  OTEL numeric severities this field is redundant and can be reproduced via reverse mapping. For formats that have levels that for some reason cannot be mapped completely to OTEL, recording the original severity text is only useful for reverse mapping - perhaps it can be stored in a k/v field and only on request (frankly, the 10-based Python levels look quite insane to me, can anyone really care about the difference between levels 36 and 37?) If the reason for a top level field for these unmapped values is format efficiency, then (a) I would say it should only be used in those cases, not in cases when mapping is unambiguous already, and (b) I would name it something else than `SeverityText`, which looks too inviting to be used without necessity.

Reply:
>For log formats that unambiguously map to OTEL numeric severities this field is redundant and can be reproduced via reverse mapping.

This is true, but it may be very difficult to do in practice. It requires the *knowledge* about reverse mapping to be present at the place where the text is necessary for displaying (i.e. in the backend). This is a pretty significant requirement which may be very difficult to satisfy in a complex collection environment. Essentially the backend needs to know about reverse mappings for all possible log sources **and** it must know which mapping to apply to each particular record (how?).

Storing `SeverityText` in the log record during the mapping process eliminates this requirement and allows the backends to display the severity text without having this reverse mapping knowledge.

Reply:
Is it a requirement that the UI display the original severity instead of only understanding the normalized values? The point of normalization is to treat all logs similarly. Otherwise the query UI gets very gnarly.

Reply:
I think it is useful to be able to show the original severity in addition to the normalized one. Preserving the original enables such capability easily. If we don't preserve it then it is  difficult to show the original (as you correctly said, we can do reverse mapping, but it is not trivial to do). 

Even if we ignore the UI possibilities I still think we should aim to preserve the original data or unambiguous representation of it, otherwise it is not possible to fullfill this general design goal:
>Translating log data from an arbitrary log format to this Data Model
> and back should ideally result in identical data.

This is an important requirement for intermediary nodes (Collectors).